File: Assignment4/scripts/verify.go
package main

// verify.go - Standalone word counter to verify MapReduce results
// Usage: go run scripts/verify.go <text-file-path-or-s3-url>
// This reads the text file directly and counts words in a single pass,
// then compares against the MapReduce JSON output.

import (
	"encoding/json"
	"fmt"
	"os"
	"regexp"
	"strings"
)

var wordRegex = regexp.MustCompile(`[^a-zA-Z]+`)

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: go run scripts/verify.go <text-file> [mapreduce-result.json]")
		os.Exit(1)
	}

	textFile := os.Args[1]

	// Read the text file
	data, err := os.ReadFile(textFile)
	if err != nil {
		fmt.Printf("Error reading text file: %v\n", err)
		os.Exit(1)
	}

	// Count words (same logic as mapper)
	expected := countWords(string(data))

	fmt.Printf("Direct word count from '%s':\n", textFile)
	fmt.Printf("  Unique words: %d\n", len(expected))
	totalWords := 0
	for _, c := range expected {
		totalWords += c
	}
	fmt.Printf("  Total words:  %d\n", totalWords)
	fmt.Println()

	// Print top 20 words
	fmt.Println("Top 20 words:")
	type wordCount struct {
		Word  string
		Count int
	}
	var sorted []wordCount
	for w, c := range expected {
		sorted = append(sorted, wordCount{w, c})
	}
	// Simple bubble sort for top 20
	for i := 0; i < len(sorted); i++ {
		for j := i + 1; j < len(sorted); j++ {
			if sorted[j].Count > sorted[i].Count {
				sorted[i], sorted[j] = sorted[j], sorted[i]
			}
		}
	}
	for i := 0; i < 20 && i < len(sorted); i++ {
		fmt.Printf("  %-20s %d\n", sorted[i].Word, sorted[i].Count)
	}

	// If a MapReduce result JSON is provided, compare
	if len(os.Args) >= 3 {
		resultFile := os.Args[2]
		resultData, err := os.ReadFile(resultFile)
		if err != nil {
			fmt.Printf("\nError reading result file: %v\n", err)
			os.Exit(1)
		}

		var actual map[string]int
		if err := json.Unmarshal(resultData, &actual); err != nil {
			fmt.Printf("\nError parsing result JSON: %v\n", err)
			os.Exit(1)
		}

		fmt.Printf("\nMapReduce result from '%s':\n", resultFile)
		fmt.Printf("  Unique words: %d\n", len(actual))
		totalActual := 0
		for _, c := range actual {
			totalActual += c
		}
		fmt.Printf("  Total words:  %d\n", totalActual)

		// Compare
		fmt.Println("\nVerification:")
		mismatches := 0
		for word, expectedCount := range expected {
			actualCount, exists := actual[word]
			if !exists {
				fmt.Printf("  MISSING: '%s' (expected %d)\n", word, expectedCount)
				mismatches++
			} else if actualCount != expectedCount {
				fmt.Printf("  MISMATCH: '%s' expected=%d actual=%d\n", word, expectedCount, actualCount)
				mismatches++
			}
		}
		for word := range actual {
			if _, exists := expected[word]; !exists {
				fmt.Printf("  EXTRA: '%s' (count=%d)\n", word, actual[word])
				mismatches++
			}
		}

		if mismatches == 0 {
			fmt.Println("  ✓ PERFECT MATCH! MapReduce result is correct.")
		} else {
			fmt.Printf("  ✗ Found %d mismatches\n", mismatches)
		}
	}
}

func countWords(text string) map[string]int {
	counts := make(map[string]int)
	words := strings.Fields(text)
	for _, word := range words {
		cleaned := strings.ToLower(wordRegex.ReplaceAllString(word, ""))
		if cleaned == "" {
			continue
		}
		counts[cleaned]++
	}
	return counts
}

File: Assignment4/scripts/verify.py
"""
Verify MapReduce results by comparing against a direct single-pass word count.

Usage:
  # Download the files from S3 first:
  aws s3 cp s3://local-map-reducer/input/sample.txt sample.txt
  aws s3 cp s3://local-map-reducer/results/final_wordcount_XXXXX.json result.json

  # Run verification:
  python scripts/verify.py sample.txt result.json
"""

import sys
import json
import re


def count_words(text):
    """Count words using the same logic as the mapper:
    - Split on whitespace
    - Lowercase
    - Strip non-alpha characters
    - Skip empty strings
    """
    counts = {}
    for word in text.split():
        cleaned = re.sub(r'[^a-zA-Z]', '', word).lower()
        if cleaned:
            counts[cleaned] = counts.get(cleaned, 0) + 1
    return counts


def main():
    if len(sys.argv) < 3:
        print("Usage: python scripts/verify.py <text-file> <mapreduce-result.json>")
        sys.exit(1)

    text_file = sys.argv[1]
    result_file = sys.argv[2]

    # Direct word count
    with open(text_file, "r") as f:
        text = f.read()
    expected = count_words(text)

    # MapReduce result
    with open(result_file, "r") as f:
        actual = json.load(f)

    print(f"Direct count:    {len(expected)} unique words, {sum(expected.values())} total")
    print(f"MapReduce count: {len(actual)} unique words, {sum(actual.values())} total")
    print()

    # Compare
    mismatches = 0

    for word, exp_count in expected.items():
        act_count = actual.get(word, 0)
        if act_count != exp_count:
            print(f"  MISMATCH: '{word}' expected={exp_count} actual={act_count}")
            mismatches += 1

    for word in actual:
        if word not in expected:
            print(f"  EXTRA:    '{word}' count={actual[word]} (not in direct count)")
            mismatches += 1

    print()
    if mismatches == 0:
        print("RESULT: PERFECT MATCH! MapReduce output is correct.")
    else:
        print(f"RESULT: {mismatches} mismatches found.")

    # Show top 10 words
    print()
    print("Top 10 words:")
    sorted_words = sorted(expected.items(), key=lambda x: x[1], reverse=True)
    for word, count in sorted_words[:10]:
        mr_count = actual.get(word, 0)
        match = "✓" if mr_count == count else "✗"
        print(f"  {match} {word:20s} direct={count:5d}  mapreduce={mr_count:5d}")


if __name__ == "__main__":
    main()
